File: src/solutions/day18.js
class Node {
  constructor(str) {
    this.str = str;
    this.value = parseInt(str);
  }

  isComma() {
    return (this.str = ",");
  }

  isOpen() {
    return this.str == "[";
  }

  isClose() {
    return this.str == "]";
  }

  isNumber() {
    return !isNaN(this.value);
  }

  static from(string) {
    return string.split("").map((ch) => new Node(ch));
  }

  static toString(nodes) {
    return nodes.map((n) => n.str).join("");
  }
}

function debug(stream) {
  for (let i = 0; i < stream.length; i++) {
    console.log(i, stream[i]);
  }
}

function explode(stream) {
  // debug(stream);
  let start = -1;
  let sum = 0;
  for (let i = 0; i < stream.length; i++) {
    let node = stream[i];
    if (node.isOpen()) {
      sum++;
    }

    if (node.isClose()) {
      sum--;
    }

    if (sum >= 5) {
      start = i;
      break;
    }
  }

  if (start == -1) {
    return stream;
  }

  let v1 = stream[start + 1].value;
  let v2 = stream[start + 3].value;

  let tmp = [new Node("0")];

  let added = false;
  let index = start - 1;
  while (index >= 0) {
    if (stream[index].isNumber() && !added) {
      tmp.unshift(new Node(`${stream[index].value + v1}`));
      added = true;
    } else {
      tmp.unshift(stream[index]);
    }
    index--;
  }

  index = start + 5;
  added = false;
  while (index < stream.length) {
    if (stream[index].isNumber() && !added) {
      tmp.push(new Node(`${stream[index].value + v2}`));
      added = true;
    } else {
      tmp.push(stream[index]);
    }
    index++;
  }

  return tmp;
}

function split(stream) {
  let tmp = [];
  let splitted = false;
  for (let i = 0; i < stream.length; i++) {
    if (!stream[i].isNumber() || splitted || stream[i].value < 10) {
      tmp.push(stream[i]);
      continue;
    }

    splitted = true;

    let a = Math.floor(stream[i].value / 2);
    let b = stream[i].value - a;

    tmp.push(
      new Node("["),
      new Node(`${a}`),
      new Node(`,`),
      new Node(`${b}`),
      new Node(`]`)
    );
  }

  return tmp;
}

function add(s1, s2) {
  return [new Node("[")]
    .concat(s1)
    .concat(new Node(","))
    .concat(s2)
    .concat(new Node("]"));
}

function reduce(node) {
  let i = 0;
  let last_s = Node.toString(node);
  let last = [...node];

  for (;;) {
    let last_s2 = Node.toString(last);
    for (;;) {
      let tmp = explode(last);
      let tmp_s = Node.toString(tmp);
      // console.log("EX:", tmp_s);

      if (tmp_s == last_s2) {
        break;
      }

      last = tmp;
      last_s2 = tmp_s;
    }

    let tmp = split(last);
    let tmp_s = Node.toString(tmp);

    // console.log("SPLIT:", tmp_s);

    if (tmp_s == last_s) {
      break;
    }

    last = tmp;
    last_s = tmp_s;
  }

  return last;
}

function magnitude(s) {
  const ex = s
    .replaceAll("[", "(")
    .replaceAll("]", ")")
    .replaceAll(",", "*3 + 2*");

  return eval(ex);
}

const data = require("fs").readFileSync(
  __dirname + "/../../inputs/day18.txt",
  "utf-8"
);

let lines = data.split("\n").map((l) => explode(Node.from(l.trim())));
let current = lines[0];

for (let i = 1; i < lines.length; i++) {
  current = add(current, lines[i]);
  current = reduce(current);
}

console.log("PART #1:", magnitude(Node.toString(current)));

const arr = [];
let max = 0;
for (let i = 0; i < lines.length; i++) {
  for (let j = 1; j < lines.length; j++) {
    let current = add(lines[i], lines[j]);
    current = reduce(current);

    arr.push(magnitude(Node.toString(current)));

    current = add(lines[j], lines[i]);
    current = reduce(current);

    let m = magnitude(Node.toString(current));
    if (m > max) {
      max = m;
    }
  }
}

console.log("PART #2:", max);

File: src/solutions/day18.rs
use std::fmt::{Debug, Display};

use itertools::Itertools;
use json::JsonValue;

use crate::day18::Item::{Close, Comma, Value};
use crate::helpers::parse_lines;
use crate::solutions::{Result, Solution};

#[derive(Default)]
pub struct DaySolution;

#[derive(Debug)]
enum Item {
    Open,
    Close,
    Comma,
    Value(usize),
}

impl Item {
    fn get_value(&self) -> usize {
        match self {
            Self::Value(value) => *value,
            _ => unreachable!(),
        }
    }
}

trait Snailfish {
    fn from_string(input: &str) -> Vec<Item>;
    fn to_string(&self) -> String;
    fn add(&mut self, other: Vec<Item>);

    fn explode(&mut self) -> bool;
    fn split(&mut self) -> bool;
    fn reduce(&mut self);
}

impl Snailfish for Vec<Item> {
    fn from_string(input: &str) -> Vec<Item> {
        input
            .chars()
            .map(|ch| match ch {
                '[' => Item::Open,
                ']' => Item::Close,
                ',' => Item::Comma,
                x => Item::Value(x.to_digit(10).unwrap_or(0) as usize),
            })
            .collect()
    }

    fn to_string(&self) -> String {
        self.iter()
            .map(|i| match i {
                Item::Open => "[".to_string(),
                Item::Close => "]".to_string(),
                Item::Comma => ",".to_string(),
                Item::Value(value) => format!("{}", value),
            })
            .join("")
    }

    fn add(&mut self, other: Vec<Item>) {
        self.insert(0, Item::Open);
        self.push(Item::Comma);
        self.extend(other);
        self.push(Item::Close)
    }

    fn explode(&mut self) -> bool {
        let mut open_count = 0;
        let mut explode_position = None;

        for (index, item) in self.iter().enumerate() {
            open_count += match item {
                Item::Open => 1,
                Item::Close => -1,
                _ => 0,
            };

            if open_count == 5 {
                explode_position = Some(index);
                break;
            }
        }

        if explode_position.is_none() {
            return false;
        }

        let explode_position = explode_position.unwrap();

        let left = self[explode_position + 1].get_value();
        let right = self[explode_position + 3].get_value();

        self[explode_position] = Item::Value(0);

        for _ in 0..4 {
            self.remove(explode_position + 1);
        }

        let mut index = explode_position - 1;
        loop {
            if let Item::Value(value) = self[index] {
                self[index] = Item::Value(value + left);
                break;
            }

            if index == 0 {
                break;
            }
            index -= 1;
        }

        index = explode_position + 1;

        while index < self.len() {
            if let Item::Value(value) = self[index] {
                self[index] = Item::Value(value + right);
                break;
            }
            index += 1;
        }

        true
    }

    fn split(&mut self) -> bool {
        if let Some((index, item)) = self
            .iter()
            .find_position(|&item| matches!(item, Item::Value(v) if *v > 9))
        {
            let value = item.get_value();
            let a = value / 2;
            let b = value - a;

            self[index] = Item::Open;
            self.insert(index + 1, Value(a));
            self.insert(index + 2, Comma);
            self.insert(index + 3, Value(b));
            self.insert(index + 4, Close);

            return true;
        }

        false
    }

    fn reduce(&mut self) {
        let mut same = false;
        while !same {
            same = true;
            while self.explode() {
                same = false;
            }

            if self.split() {
                same = false;
            }
        }
    }
}

impl DaySolution {
    fn magnitude(&self, value: JsonValue) -> f64 {
        match value {
            JsonValue::Array(arr) => {
                let a = arr[0].clone();
                let b = arr[1].clone();

                match (a, b) {
                    (JsonValue::Number(a), JsonValue::Number(b)) => {
                        let a: f64 = a.into();
                        let b: f64 = b.into();
                        a * 3.0 + b * 2.0
                    }
                    (a, JsonValue::Number(b)) => {
                        let b: f64 = b.into();
                        self.magnitude(a) + b * 2.0
                    }
                    (JsonValue::Number(a), b) => {
                        let a: f64 = a.into();
                        a * 3.0 + self.magnitude(b)
                    }
                    (a, b) => self.magnitude(a) * 3.0 + self.magnitude(b) * 2.0,
                }
            }
            _ => 0.0,
        }
    }
}

impl Solution for DaySolution {
    fn part_1(&mut self, input: Option<String>) -> Result<Box<dyn Display>> {
        let lines = parse_lines::<String>(input);
        let mut current = Vec::<Item>::from_string(&lines[0]);

        for line in lines.iter().skip(1) {
            current.add(Vec::from_string(line));
            current.reduce();
        }

        let value = json::parse(&current.to_string()).unwrap();

        Ok(Box::new(self.magnitude(value) as usize))
    }

    fn part_2(&mut self, input: Option<String>) -> Result<Box<dyn Display>> {
        let lines = parse_lines::<String>(input);

        let max = lines
            .iter()
            .permutations(2)
            .map(|pair| {
                let (a, b) = (pair[0], pair[1]);
                let mut a = Vec::from_string(a);

                a.add(Vec::from_string(b));
                a.reduce();

                let json = json::parse(&a.to_string()).unwrap();
                self.magnitude(json) as usize
            })
            .max()
            .unwrap();

        Ok(Box::new(max))
    }
}

#[cfg(test)]
mod tests {
    use crate::day18::DaySolution;
    use crate::Solution;

    #[test]
    fn part_1() {
        let input = include_str!("../../inputs/day18_demo.txt");
        let result = DaySolution::default()
            .part_1(Some(input.to_string()))
            .unwrap();

        assert_eq!("4140", result.to_string());

        let input = "[[[[8,7],[7,7]],[[8,6],[7,7]]],[[[0,7],[6,6]],[8,7]]]";

        let result = DaySolution::default()
            .part_1(Some(input.to_string()))
            .unwrap();

        assert_eq!("3488", result.to_string());
    }

    #[test]
    fn part_2() {
        let input = include_str!("../../inputs/day18_demo.txt");
        let result = DaySolution::default()
            .part_2(Some(input.to_string()))
            .unwrap();

        assert_eq!("3993", result.to_string());
    }
}
